File: Data-structure/BST.php
<?php
 error_reporting(E_ALL);
class Node {
	public $data;
	public $parent;
	public $left;
	public $right;

	public function __construct($data) {
		$this->data = $data;
		$this->parent = null;
		$this->left = null;
		$this->right = null;
	}
}

class BinarySearchTree {
	public $root;

	public function __construct() {
		$this->root = null;
	}

	public function insertNode($data) {
		$newNode = new Node($data);
		$currentNode = $this->root;
		$parentNode = null;

		if($currentNode == null) {
			$this->root = $newNode;
		} else {
			while ($currentNode != null) {
				$parentNode = $currentNode;
				if($newNode->data < $currentNode->data) {
					$currentNode = $currentNode->left;
				} else {
					$currentNode = $currentNode->right;
				}
			}
			
			$newNode->parent = $parentNode;
			if($newNode->data < $parentNode->data) {
				$parentNode->left = $newNode;
			} else {
				$parentNode->right = $newNode;
			}
		}
	}

	public function inOrderTreeWalk($node) {
		if ($node != null) {
			$this->inOrderTreeWalk($node->left);
			echo $node->data.' ';
			$this->inOrderTreeWalk($node->right);
		}
	}

	public function preOrderTreeWalk($node) {
		if ($node != null) {
			echo $node->data.' ';
			$this->inOrderTreeWalk($node->left);
			$this->inOrderTreeWalk($node->right);
		}
	}

	public function PostOrderTreeWalk ($node) {
		if ($node != null) {
			$this->inOrderTreeWalk($node->left);
			$this->inOrderTreeWalk($node->right);
			echo $node->data.' ';
		}
	}
}

$bst = new BinarySearchTree();
$bst->insertNode(10);
$bst->insertNode(9);
$bst->insertNode(34);
echo "Inorder (Left, Root, Right): ";
$bst->inOrderTreeWalk($bst->root);
echo "<br>";
echo "Preorder (Root, Left, Right): ";
$bst->preOrderTreeWalk($bst->root);
echo "<br>";
echo "Postorder (Left, Right, Root): ";
$bst->PostOrderTreeWalk($bst->root);

File: Data-structure/BST.py
class Node(object):
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None


class BST(object):
    def __init__(self):
        self.root = None

    def insert_node(self, data):
        new_node = Node(data)
        if self.root is None:
            self.root = new_node
            return

        current_node = self.root
        while current_node is not None:
            parent_node = current_node
            if data > current_node.data:
                current_node = current_node.right
            else:
                current_node = current_node.left

        if new_node.data < parent_node.data:
            parent_node.left = new_node
        else:
            parent_node.right = new_node

    def pre_order(self, node):
        if node is not None:
            self.pre_order(node.left)
            print(node.data)
            self.pre_order(node.right)

    def in_order(self, node):
        if node is not None:
            print(node.data)
            self.in_order(node.left)
            self.in_order(node.right)

    def post_order(self, node):
        if node is not None:
            self.post_order(node.left)
            self.post_order(node.right)
            print(node.data)


b_tree = BST()
b_tree.insert_node(3)
b_tree.insert_node(4)
b_tree.insert_node(1)
b_tree.insert_node(40)
b_tree.insert_node(0)
b_tree.insert_node(2)
# b_tree.pre_order(b_tree.root)
# b_tree.in_order(b_tree.root)
# b_tree.post_order(b_tree.root)
